pkg/kubelet/kubelet_pods.go

```diff
@@ -1624,6 +1624,26 @@ func (kl *Kubelet) generateAPIPodStatus(pod *v1.Pod, podStatus *kubecontainer.Po
 		if err != nil {
 			klog.V(4).InfoS("Cannot get host IPs", "err", err)
 		} else {
+			if s.HostIP != "" {
+				ok, err := utilnet.IsDualStackIPStrings([]string{s.HostIP, hostIPs[0].String()})
+				if err != nil {
+					kl.recorder.Eventf(pod, v1.EventTypeWarning, "InvalidHostIPs", "host IPs are invalid %q or %q", s.HostIP, hostIPs[0].String())
+				}
+				if ok {
+					// Once CloudDualStackNodeIPs is beta, <EXPLICIT-IPV4-ADDRESS>,<EXPLICIT-IPV6-ADDRESS> can be set to --node-ip.
+					if utilnet.IPFamilyOf(hostIPs[0]) == utilnet.IPv4 {
+						// kubelet auto-detected an IPv4 node IP (hostIPs[0]), but the cloud provider wants this to be
+						// a single-stack IPv6 or dual-stack IPv6-primary node
+						kl.recorder.Eventf(pod, v1.EventTypeWarning, "HostIPsIPFamilyMismatch",
+							"set [::] or an explicit IPv4 address to --node-ip")
+					} else {
+						// kubelet auto-detected an IPv6 node IP (hostIPs[0]), but the cloud provider wants this to be
+						// a single-stack IPv4 or dual-stack IPv4-primary node
+						kl.recorder.Eventf(pod, v1.EventTypeWarning, "HostIPsIPFamilyMismatch",
+							"do not set --node-ip or set an explicit IPv4 address to --node-ip")
```

Comment:
likewise. Also, the fix here is to pass an explicit _IPv6_ address, not IPv4.

actually, we don't really even need separate errors for the v4 and v6 case. Just replace this all (1627-1646) with:

```
if s.HostIP != "" && utilnet.IPFamilyOfString(s.HostIP) != utilnet.IPFamilyOf(hostIPs[0]) {
        kl.recorder.Eventf(pod, v1.EventTypeWarning, "HostIPsIPFamilyMismatch",
                "Kubelet detected an IPv%s node IP (%s), but the cloud provider selected an IPv%s node IP (%s); pass an explicit `--node-ip` to kubelet to fix this.",
                utilnet.IPFamilyOfString(s.HostIP), s.HostIP, utilnet.IPFamilyOf(hostIPs[0])), hostIPs[0].String())
}
```

Reply:
done